test/quantization/test_quantize_script.py

```diff
@@ -2348,10 +2339,81 @@ def forward(self, x):
 
-        model = quantize_dynamic_script(model, qconfig_dict, [data])
+        model = quantize_dynamic_script(model, qconfig_dict)
         FileCheck().check("quantized::linear_dynamic") \
                    .check_not("aten::_choose_qparams_per_tensor") \
                    .run(model.graph)
+
+    @override_qengines
+    def test_dynamic_shared_weights(self):
+        class myMod(torch.nn.Module):
+            def __init__(self, weight):
+                super().__init__()
+                self.linear = nn.Linear(5, 5)
+                self.linear.weight = weight
+
+            def forward(self, x):
+                return self.linear(x)
+
+        class DynamicModel(torch.nn.Module):
+            def __init__(self):
+                super(DynamicModel, self).__init__()
+                self.weight = torch.nn.Parameter(torch.ones(5, 5))
+                self.mod1 = myMod(self.weight)
+
+            def forward(self, x):
+                y = self.mod1(x)
+                z = torch.nn.functional.linear(y, self.weight)
+                return z
+
+        model = torch.jit.script(DynamicModel()).eval()
+        data = torch.randn(5, 5, dtype=torch.float)
+        for op in ['mod1', '']:
+            qconfig_dict = {op: default_dynamic_qconfig}
+            model_graph = quantize_dynamic_script(model, qconfig_dict)
+            out_graph = model_graph(data)
+
+            # Explicitly call forward on model before convert
+            m = prepare_dynamic_script(model, qconfig_dict)
+            m(data)
+            m = convert_dynamic_script(m, debug=False)
+            out_ref = model_graph(data)
+            self.assertEqual(out_graph, out_ref)
+
+    @override_qengines
+    def test_dynamic_with_if(self):
+        class Res(torch.nn.Module):
+            def __init__(self):
+                super(Res, self).__init__()
+                self.weight = torch.nn.Parameter(torch.ones(5, 5))
+
+            def forward(self, x, cond):
+                # type: (Tensor, bool) -> Tensor
+                if cond:
+                    return torch.nn.functional.linear(x, self.weight)
+                else:
+                    return torch.nn.functional.linear(x, self.weight)
+
+        class M(torch.nn.Module):
+            def __init__(self):
+                super(M, self).__init__()
+                self.res1 = Res()
+                self.res2 = Res()
+
+            def forward(self, x):
+                x = self.res1(x, True)
+                x = self.res2(x, False)
+                return x
+
+        model = torch.jit.script(M()).eval()
+        data = torch.randn(5, 5, dtype=torch.float)
+        qconfig_dict = {'': default_dynamic_qconfig}
+        m1 = quantize_dynamic_script(model, qconfig_dict)
+        out_graph = m1(data)
+
+        m2 = prepare_dynamic_script(model, qconfig_dict)
+        m2 = convert_dynamic_script(m2)
+        out_ref = m2(data)
+        self.assertEqual(out_graph, out_ref)
```

Comment:
We need to explicitly check the graph of the dynamic quantized model, checking for consistency between prepare->convert and quantize is not sufficient as it will not catch errors made in both paths.

Reply:
We are checking numerics here to make sure both ways work. The graph will not change since the extra weight subgraphs aren't added to the original graph.

Reply:
This is a general issue, but can be addressed in a later PR to check if we handle shared weights correctly